haystack/generator/seq2seq.py

```diff
@@ -0,0 +1,204 @@
+import logging
+from abc import abstractmethod
+from typing import Dict, List, Optional
+
+import torch
+from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
+
+from haystack import Document
+from haystack.generator.base import BaseGenerator
+from haystack.retriever.base import BaseRetriever
+from haystack.retriever.dense import EmbeddingRetriever
+
+logger = logging.getLogger(__name__)
+
+
+class Seq2SeqGenerator(BaseGenerator):
+    """
+        A generic sequence-to-sequence generator based on HuggingFace. Subclasses should
+        implement prepare_model_input abstract method
+
+        Text generation is supported by so called auto-regressive language models like GPT2,
+        XLNet, XLM, Bart, T5 and others. In fact, any HuggingFace language model that extends
+        GenerationMixin can be used by Seq2SeqGenerator
+
+        See https://huggingface.co/transformers/main_classes/model.html?transformers.generation_utils.GenerationMixin#transformers.generation_utils.GenerationMixin
+        as well as https://huggingface.co/blog/how-to-generate
+
+        For a list of all text-generation models see https://huggingface.co/models?pipeline_tag=text-generation
+
+        **Example**
+
+        ```python
+        |     query = "Why is Dothraki language important?"
+        |
+        |     # Retrieve related documents from retriever
+        |     retrieved_docs = retriever.retrieve(query=query)
+        |
+        |     # Now generate answer from query and retrieved documents
+        |     generator.predict(
+        |        query=query,
+        |        documents=retrieved_docs,
+        |        top_k=1
+        |     )
+        |
+        |     # Answer
+        |
+        |     {'answers': [" The Dothraki language is a constructed fictional language. It's important because George R.R. Martin wrote it."],
+        |      'query': 'Why is Dothraki language important?'}
+        |
+        ```
+    """
+
+    def __init__(
+            self,
+            model_name_or_path: str,
+            retriever: Optional[BaseRetriever] = None,
```

Comment:
How would we use a retriever here? From my understanding we just condition our generation on texts (and not the embeddings from an external retriever as in RAG)

Reply:
Hey @tholor 

Thanks for the praise. Most likely it is a fresh set of eyes looking at the problem :-)

> Let's tackle this before diving into a low-level review.

I agree with you. Let's find the best broad design strokes before diving into details.

> Is that also your understanding or do I miss more specialties here? Having clarity here will help us to settle on the design.

You are right on the spot. The key method interfaces seem to be embedding queries/passages. As Yacine's Retribert also distinguishes encoding queries and passage into different projects (using different methods as well) it seemed natural to isolate these high-level signatures. 

>  Having one class per model architecture could easily end up in a mess. That's why we don't have a RobertaReader, BERTReader etc.

I agree here except that I don't not foresee explosion of new EmbeddingRetriever subclasses/implementations as the current default implementation covers a lot of LM use cases, SentenceTransformer covers its use case, and finally Retribert its own. I am open to new suggestions here, let's find the best design approach. 

>A bit related to the above design questions: Is there a meaningful way to avoid the very model and dataset specific BartEli5Generator? I I think it will become eventually very messy to have one class per model. I can see that the only custom part here is the preparation of the input documents with < P>. 

You are right, very likely it will, but I am not sure how to work around it? I foresee the model input preparation to be the only model specific part of the code - think T5 task directives, BART placeholders etc etc, while everything else stays the same (the generate method invocation and parameters from GenerationMixin)

> We probably want to move Seq2seq to generators/transformers.py as the modules are currently rather divided by framework than by architecture and the RAG implementation is also in generators/transformers.py

No problem, will do, good decision.

Reply:
Yes, no need, a total oversight on my behalf here.